### PR TITLE
Support for MONC model grid.

### PR DESCRIPTION
**Pull request by ReadingClouds**
*Support for MONC*
Primarily changes to support the MONC grid. This has:
- First u point at first p point + dx/2, so it makes sense to have p[0] at x=0.
- First v point at first p point + dy/2, so it makes sense to have p[0] at y=0.
- Virtual p point at z = -dz/2.

Main changes are in creating synthetic data for testing.

### CODE REVIEW

File: advtraj/utils/grid.py
Comment (leifdenby) [inline advtraj/utils/grid.py:67]: I think this makes sense, there was a bug in my code before for coordinates on cell-faces, is that right?

File: CHANGELOG.md
Comment (leifdenby) [inline CHANGELOG.md:8]: I think it would be clearer to put this in `advtraj.utils.grid` so we detail in there how the MONC grid works with coordinates on cell-faces. Also, I'm not sure what you mean by "p point" here. Does that refer to points that are for scalar values? Sorry if this is c-grid jargon that I'm just unaware of / have forgotten.

For the changelog I would just write "add support for MONC grid by implementing its c-grid configuration for where the position scalars are stored" or something like that, assuming that is correct.

File: tests/utils.py
Comment (leifdenby) [inline tests/utils.py:111]: good catch!

File: tests/utils.py
Comment (leifdenby) [inline tests/utils.py:130]: I guess this isn't needed anymore because we explicitly indicate whether the domain is xy-periodic when calling `advtraj.integrate_trajectories`, is that why you're removing this? I think I must've added this when I was developing the code and didn't realise while changing things around that it wasn't needed anymore.

**Review by leifdenby — CHANGES_REQUESTED** [review]
Looks good to me @ReadingClouds! I'm a bit unclear about how MONC's grid is different (in terms of the position scalars), but I think I get it from your code changes. My suggestions mainly revolve about making that more explicit. I like the idea of introducing a `grid_type`, but wouldn't it be enough to just say that MONC is not cell-centered? Or said differently, is there a label that we could give this kind of grid that is more general than it's a "monc"-type grid. In case people with different models use a similar type of grid this would be useful I think. Hope these comments are useful!

File: tests/utils.py
Comment (ReadingClouds) [reply tests/utils.py:130]: Basically, except in test/utils.py above, "xy_periodic" is an attribute everywhere. (I plan to add it to my 'MONC io" grid correction code). You needed this here to convert the class attribute to a member of attr. It you set it to an attribute in the first place it gets carried over. I think it's better from a NetCDF point of view to stuff it in attrs, but I'm not very NetCDF savvy.

File: advtraj/utils/grid.py
Comment (ReadingClouds) [reply advtraj/utils/grid.py:67]: Yes

**Comment by ReadingClouds** [comment]
Re grid types, I'd welcome more general input, though a) this is a perenial problem and b) there's probably a CDF spec we should follow, but I haven't tried to find or understand it. IMV, the clearest solution is the xarray one - actually have the coords stored with the data, rather than some 'code'. It still helps, sometimes, to recognise a stencil.

Both UCLALES and MONC are C-grid/Lorenz, I believe. So everything is at the 'centre' of a 'cell' apart from u,v,w which are at centres of cell faces. I (and much documentation) label the different points u,v,w,p - p, not rho or theta because, in a Bousinesq model, rho is pretty irrelevant, and thete, on a Charney-Philips vertical grid is at w points (so the UM has rho-levels and theta-levels!).

However, being the same grid structure doesn't meen the labelling is the same. If you want to think of the volume as filled with cells, your 'cell-centred' spec makes sense. However a) MONC has the pesky extra level below the surface and b) MONC labels the first p point, the u point with x dx/2 bigger and the v point with y dy/2 bigger with the same number. 

This means that the 'boundary' point for the first cell is located at u point i=-1 (counting the python way from 0). In python, this is the last point so it doesn't make sense to me to say this is at x=0. 

It then makes sense to me have the first (i.e. labelled 0 in python) p point at  x=0, so the first u is at dx/2,0 and the first v at 0,dy/2. 

However, I can see some merit in having u[0,...] at dx and p[0,...] at dx/2. Then u[-1,...] is at L_x and the domain size is clear. I think I'll canvas some users.

**Comment by ReadingClouds** [comment]
I believe I have responded to review comments.

File: advtraj/utils/grid.py
Comment (leifdenby) [inline advtraj/utils/grid.py:6]: this is perfect! Thanks!

**Comment by leifdenby** [comment]
Looks great to me! I usually do a "squash and merge" since that way we only have a single commit of all these changes.

I'll let you do the merge! Thanks again @ReadingClouds

**Comment by ReadingClouds** [comment]
I don't believe I have write access to do the merge. Probably for the best.

Prof. Peter Clark
Department of Meteorology
University of Reading
Tel: +44(0) 118 378 7908<tel:+44%20118%20378%207908>
email: ***@***.******@***.***>

**Comment by ReadingClouds** [comment]
Thanks!

Prof. Peter Clark
Department of Meteorology
University of Reading
Tel: +44(0) 118 378 7908<tel:+44%20118%20378%207908>
email: ***@***.******@***.***>

**Comment by leifdenby** [comment]
Thanks for doing this @ReadingClouds! It's exciting that we'll be able to support more than one LES :partying_face:

**Pull request merged**
